modules/geo-layers/src/tile-layer/tile-2d-header.js

```diff
@@ -64,47 +69,56 @@ export default class Tile2DHeader {
       return;
     }
 
-    let tileData;
+    let tileData = null;
     let error;
     try {
-      tileData = await getTileData({x, y, z, bbox, signal});
+      tileData = await getData({x, y, z, bbox, signal});
     } catch (err) {
       error = err || true;
     } finally {
       requestToken.done();
-
-      if (this._isCancelled && !tileData) {
-        this._isLoaded = false;
-      } else {
-        // Consider it loaded if we tried to cancel but `getTileData` still returned data
-        this._isLoaded = true;
-        this._isCancelled = false;
-      }
     }
 
-    if (!this._isLoaded) {
+    // If loadData has been called with a newer version, discard the result from this operation
+    if (loaderId !== this._loaderId) {
+      return;
+    }
+    // Clear the `isLoading` flag
+    this._loader = undefined;
+    // Rewrite tile content with the result of getTileData if successful, or `null` in case of
+    // error or cancellation
+    this.content = tileData;
+    // If cancelled, do not invoke the callbacks
+    // Consider it loaded if we tried to cancel but `getTileData` still returned data
+    if (this._isCancelled && !tileData) {
+      this._isLoaded = false;
       return;
     }
+    this._isLoaded = true;
+    this._isCancelled = false;
 
     if (error) {
-      this.onTileError(error, this);
+      onError(error, this);
     } else {
-      this.content = tileData;
-      this.onTileLoad(this);
+      onLoad(this);
     }
   }
   /* eslint-enable max-statements */
 
-  loadData(getTileData, requestScheduler) {
-    if (!getTileData) {
-      return;
-    }
-
+  loadData(opts) {
     this._isCancelled = false;
-    this._loader = this._loadData(getTileData, requestScheduler);
-    this._loader.finally(() => {
+    this._needsReload = false;
+    this._loaderId++;
+    this._loader = this._loadData(opts);
```

Comment:
```suggestion
    this._isLoaded = false;
    this._needsReload = false;
    this._loaderId++;
    this._loader = this._loadData(opts);
```

I think?

Since the state of loading upstream depends on all tiles being loaded, I think when the update triggers for the `getTileData` change, the tile will remain loaded when this method is called again since `isLoaded` has never been set to `false`.  For example. I think the `onViewportLoad` method will be called immediately when `getTileData` changes not when the new data has loaded.

Reply:
I left the `isLoaded` flag on because it is used in `renderLayers` to determine if the tile should be rendered. But you are right that it doesn't work for the `onViewportLoad` callback. I think it's right to reset it here and we need a different check for layer rendering.